### PR TITLE
Manchester Class NW5 - Anvar Azizi - JAVASCRIPT CORE 1 - Week 1

### PR DESCRIPTION
**Pull request by rebwar1**
**Volunteers: Are you marking this coursework?** _You can find a guide on how to mark this coursework in `HOW_TO_MARK.md` in the root of this repository_

# Your Details

- Your Name: Anvar 
- Your City: Manchester
- Your Slack Name: rebwar

# Homework Details

- Module: JavaScript Core 1
- Week: 1

# Notes

- What did you find easy?

- What did you find hard?

- What do you still not understand?

- Any other notes?

### CODE REVIEW

File: exercises/D-strings/exercise.js [hunk 1]
```diff
@@ -1,3 +1,8 @@
 // Start by creating a variable `message`
-
+function messageType(massage) {
+  return `This is a ${typeof massage} and type of ${typeof massage} is: ${typeof massage}`;
+}
+const message = messageType("Rebwar");
 console.log(message);
+console.log(messageType(5));
+console.log(messageType({}));
```
Comment (KarenPudner) [inline exercises/D-strings/exercise.js:8]: Good to see you experimenting with other types, not just string.

File: exercises/F-strings-methods/exercise2.js [hunk 2]
```diff
@@ -1,3 +1,14 @@
...
+function lengthRegEx(name) {
+  let trimeStr = name.replace(/\s/g, "");
+  return `my name is ${name} and my name is ${trimeStr.length} characters long`;
+}
+const nametrim = " Rebwar a  ";
+const nameReqEx = " Rebwar a  ";
+const messagetrim = lengthTrim(nametrim);
+const messageRegEx = lengthRegEx(nameReqEx);
+console.log(`length with trim(): ${messagetrim}`);
+console.log(`length with RegExp: ${messageRegEx}`);
```
Comment (KarenPudner) [inline exercises/F-strings-methods/exercise2.js:14]: Good use of trim and also regex, which has the advantage of getting rid of the space between words.

File: extra/1-currency-conversion.js [hunk 3]
```diff
@@ -15,7 +18,11 @@ function convertToUSD() {}
   They have also decided that they should add a 1% fee to all foreign transactions, which means you only convert 99% of the £ to BRL.
 */
 
-function convertToBRL() {}
+function convertToBRL(pound) {
+  const fee = pound / 100;
+  const exchange = (pound - fee) * 5.7;
+  return +exchange.toFixed(2);
```
Comment (KarenPudner) [inline extra/1-currency-conversion.js:24]: Is there a reason why you have used a unary plus sign here?

File: extra/2-piping.js [hunk 4]
```diff
@@ -16,26 +16,28 @@
...
 
 // Why can this code be seen as bad practice? Comment your answer.
-let badCode =
+// let badCode = format(startingValue);
+let badCode = format(startingValue);
 
 /* BETTER PRACTICE */
 
-let goodCode =
+let goodCode = format(startingValue);
```
Comment (KarenPudner) [inline extra/2-piping.js:40]: You should try and call the 3 functions that you have created here, as the good code example.

File: extra/3-magic-8-ball.js [hunk 5]
```diff
@@ -101,7 +114,9 @@ test("magic 8 ball returns different values each time", () => {
     );
   }
 
-  let seenPositivities = new Set(Array.from(seenAnswers.values()).map(checkAnswer));
+  let seenPositivities = new Set(
+    Array.from(seenAnswers.values()).map(checkAnswer)
+  );
```
Comment (KarenPudner) [inline extra/3-magic-8-ball.js:119]: There isn't an elegant way to do this - one option is to put all the possible answers into 1 array in order, get one at random, and the use the index to decide whether it is very positive, positive etc. Once you learn about objects in a future lesson, you could find a better way of doing this.

File: mandatory/4-tax.js [hunk 6]
```diff
@@ -17,7 +20,11 @@ function calculateSalesTax() {}
   Remember that the prices must include the sales tax (hint: you already wrote a function for this!)
 */
 
-function addTaxAndFormatCurrency() {}
+function addTaxAndFormatCurrency(price) {
+  let taxOf = (price * 20) / 100;
+  let total = taxOf + price;
+  return `£${total.toFixed(2)}`;
+}
```
Comment (KarenPudner) [inline mandatory/4-tax.js:27]: Good use of .toFixed

File: mandatory/3-function-output.js [hunk 7]
```diff
@@ -1,16 +1,21 @@
 // Add comments to explain what this function does. You're meant to use Google!
+
 function getRandomNumber() {
   return Math.random() * 10;
+  //random returns with 10 will result in a max value of 9.999
```
Comment (KarenPudner) [inline mandatory/3-function-output.js:5]: You should mention that the lowest value is 0.

**Review by KarenPudner — APPROVED** [review]
Good work - you have answered all the questions and your tests all pass, except for magic 8 ball. This is a difficult exercise to do elegantly, but there are some possible solutions - see my comment. But the rest of your work is very good. You have clearly understood the concepts covered, given your variables meaningful names and tried to find more than one solution in some cases.

**Comment by rebwar1** [comment]
> magic 8 ball

**Comment by rebwar1** [comment]
> Good work - you have answered all the questions and your tests all pass, except for magic 8 ball. This is a difficult exercise to do elegantly, but there are some possible solutions - see my comment. But the rest of your work is very good. You have clearly understood the concepts covered, given your variables meaningful names and tried to find more than one solution in some cases.

in fact I can make a magic 8 ball in scratch by my selves but really I don't understand this test

File: extra/1-currency-conversion.js (hunk 3 above)
Comment (rebwar1) [reply extra/1-currency-conversion.js:24]: it's return string so I use + sign instead Number() to change string to number

File: extra/3-magic-8-ball.js (hunk 5 above)
Comment (rebwar1) [reply extra/3-magic-8-ball.js:119]: this not my code, this is code from jest. my solution start from line 67 to 76

**Pull request closed without merging**